Adding possibility to configure multiple service accounts for different tasks

cezkuj opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

The pipeline specification allows multiple different service accounts for tasks so that different task can use the right service account.
E.g. build task may need different privileges than deploy task

Fixes
 - #777
<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md)
for more details._

# Release Notes
<!--
Does your PR contain User facing changes?

If so, briefly describe them here so we can include this description in the
release notes for the next release!
-->

```
Adding possibility to configure multiple service accounts for different tasks
```

dibyom commented:
/ok-to-test

dibyom commented on Gopkg.lock:
```diff
@@ -22,7 +22,7 @@
 
 [[projects]]
-  digest = "1:355d491068cd888a3725d68cd09681a49251d6d8e3cb49120bf6b83a86c7ee0f"
+  digest = "1:c7ec3a8daf56c9d7c4bb36f498f6db3f8c339b1142300dcd4328950574f91c6f"
```
Was updating this file intentional?

cezkuj replied:
This file was updated by hack/update-deps.sh (dep ensure). First version of PR didn't have that change and failed on integration test(please see Prow history).

cezkuj commented:
/assign @dlorenc

dlorenc commented:
In general this makes sense to me, but given it's an API change we should probably have another reviewer look.

My reasoning for why this would make sense:

- Tasks within a pipeline may require access to different sets of resources, requiring different credentials. The principle of least-privilege would ask that Tasks get the minimum access they need, rather than the "highest-common-denominator".
- Some Tasks may be authored in-house while others are reused from something like our catalog. Users may feel comfortable offering higher levels of access to tasks that they authored.

dlorenc commented:
cc @bobcatfish @ImJasonH to give the second sign off for an API review

vdemeester commented:
Same as @dlorenc, make sense to me ; mainly though of the first point, aka least-privilege :angel:

bobcatfish commented:
Thanks for including the docs in the PR @cezkuj and welcome to the project!!! :heart:

> In general this makes sense to me, but given it's an API change we should probably have another reviewer look.

I agree in general with being able to specify service accounts at the level of the Tasks within a Pipeline (and possibly also at the level of the PipelineResources, but that's another story :D).

**However** I don't think we should put this in the `Pipeline` though: right now service accounts are considered runtime information and are included in the `PipelineRun` and `TaskRun` specs - I think if we want to add task specific service accounts, we should do that in the `PipelineRun`. A couple ideas how we could do that:

We could have a section for config specific to each `task`:

```yaml
kind: PipelineRun
...
spec:
  pipelineRef:
    name: demo-pipeline
  serviceAccount: 'default'
  tasks:
  ...
  - name: source-repo
    serviceAccount: my-special-service-account
```

Or we could have a serviceAccount section:

```yaml
kind: PipelineRun
...
spec:
  pipelineRef:
    name: demo-pipeline
  serviceAccount: 'default'
  serviceAccounts:
  ...
  - taskName: source-repo
    serviceAccount: my-special-service-account
```

What do you think @vdemeester @dlorenc @ImJasonH  ?

bobcatfish commented:
/hold

bobcatfish commented:
(Adding this to the WG agenda tomorrow in case folks want to discuss in person)

vdemeester commented:
oh yeah, definitely on `PipelineRun` :+1: 
@bobcatfish one question on your 2nd examples : having `serviceAccounts` would allow us to have multiple service accounts on each tasks ?

bobcatfish commented:
> @bobcatfish one question on your 2nd examples : having serviceAccounts would allow us to have multiple service accounts on each tasks ?

Hm i don't think so - I'm not sure if that would work with our current design :thinking: 

The only difference b/w the two examples is whether the top level thing is called `tasks` and each section is about configuration for `tasks` (which would assume we would want to put other `task` level info here later? maybe planning too far ahead, not sure what that would be ... :sweat_smile: ) and the other example has the top level thing called `serviceAccounts` instead and is just about serviceAccounts <-- probably simpler

cezkuj commented:
Hmm, I see some reasoning behind moving it to `PipelineRun`. I would definitely go with `serviceAccounts` way as it seems more obvious for me. I will try to join WG tomorrow in case of further discussion.

bobcatfish commented:
Didn't end up having time to discuss in the WG - I think go ahead with the `serviceAccounts` design or something similar @cezkuj ! If you want more discussion you could create a design doc, but I think it's a small enough change that we could just continue to discuss in the PR.

cezkuj commented:
Updated PR with `serviceAccounts` implementation. Wasn't sure if `taskName` in `serviceAccounts` should refer to `Task` or `PipelineTask` - I've chosen `Task`. As a side effect of update-deps.sh LICENSE got also updated - not sure if this is wanted.

bobcatfish commented on docs/pipelineruns.md:
````diff
@@ -94,6 +94,15 @@ that is in the
+  serviceAccounts:
+    - taskName: build-task
+      serviceAccount: sa-for-build
+```
````
this interface looks great! :+1:

bobcatfish commented on docs/pipelineruns.md:
```diff
@@ -94,6 +94,15 @@ that is in the
 [namespace](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/)
 of the `TaskRun` resource object.
 
+There is also possibility to configure service account for concrete `Task`, overriding `ServiceAccount` set in `PipelineRun`, for example:
```
i think it would be good to be clear that the `taskName` below is the name of the task within the Pipeline, vs. the `Task` crd.

maybe you could just add a bit more to the example, e.g. something like:


For example with this `serviceAccount` configuration:
```yaml
spec:
  serviceAccount: sa-1
  serviceAccounts:
    - taskName: build-task
      serviceAccount: sa-for-build
```

If used with this `Pipeline`, `test-task` will use the `ServiceAccount` `sa-1`, while `build-task` will use `sa-for-build`.

```yaml
kind: Pipeline
spec:
  tasks:
    - name: build-task
      taskRef:
        name: build-push
  tasks:
    - name: test-task
      taskRef:
        name: test
```

cezkuj replied:
Hmm, TBH, as far as I see I've implemented it to refer to `Task` crd. Now it seems less obvious to me why I'd go this way (plus it seems obvious for you to go with referring `PipelineTask` ). I'll change that and update docs as you've mentioned. Thanks for review!:)

cezkuj replied:
Changed to `PipelineTask` way and extended docs. Please take a look once again. :)

bobcatfish commented on test/builder/pipeline.go:
```diff
@@ -310,6 +310,17 @@ func PipelineRunServiceAccount(sa string) PipelineRunSpecOp {
 }
 
+// PipelineRunServiceAccountTask overrides the service account for given Task in PipelineRun.
+func PipelineRunServiceAccountTask(taskName, sa string) PipelineRunSpecOp {
```
the functions in this file also have their own tests in https://github.com/tektoncd/pipeline/blob/master/test/builder/pipeline_test.go could you add some coverage for this function there also? :D

cezkuj replied:
Done.

cezkuj commented:
/retest

bobcatfish commented on docs/pipelineruns.md:
```diff
@@ -98,6 +100,34 @@ of the `TaskRun` resource object.
+  tasks:
+    - name: test-task
+      taskRef:
+        name: test
```
👍 thanks for adding this!
